src/Microsoft.Diagnostics.DebugServices.Implementation/ThreadServiceFromDataReader.cs

```diff
@@ -54,7 +55,7 @@ protected override bool GetThreadContext(uint threadId, uint contextFlags, uint
 
         protected override IEnumerable<IThread> GetThreadsInner()
         {
-            return _threadReader.EnumerateOSThreadIds().Select((uint id, int index) => new Thread(this, index, id)).Cast<IThread>();
+            return _threadReader.EnumerateOSThreadIds().Select((uint id, int index) => new Thread(this, index, id)).ToImmutableArray();
```

Comment:
If we don't use something like ToImmutableArray or ToArray to actually make/return a copy of the Threads, the Select() gets evaluated as the GetThreadsInner() callers enumerate.  In this case, it isn't necessary because ThreadService.GetThreads() creates a dictionary that is saved in _threads.  I was being paranoid because I did run into a problem somewhere else (I don't remember exactly where maybe in the ModuleService) where the resulting saved snapshot (like _threads or _modules) caused the Linq functions to run so it didn't behave like a actual snapshot/copy.

It is hard to explain exactly what happened.  I might be able to explain it better verbally.

Reply:
I'll change it back.